### ISSUE TITLE
grad should always return a value for linear functions

### ISSUE THREAD

**Issue by rileyjmurray** [cvxgrp/cvxpy#459]
The current implementation of "grad" in cvxpy/atoms/atom.py returns None if the variables in the underlying expression have not been assigned values.

However if "expr" satisfies `expr.is_affine() == True` and `expr.shape == ()`, then the gradient is well-defined without specifying a value for any variables appearing in `expr`.

At present, there is a way to access the gradient. If `vars = expr.variables()` then we can set `v.value = np.zeros(v.shape)` for each `v in vars`. Then calling `expr.grad` will return the correct dictionary from variables to (vectorized) gradients.  Strictly speaking, if `expr.variables()` is an empty list, then `expr.grad` should return 0 (as a scalar).

**Comment by rileyjmurray** [comment]
It's okay that this functionality not be exposed, but for affine functions it seems to make a good deal of sense. 

If a problem is put in standard conic form, then it's completely specified by the gradient of the affine constraint expression, and the cone membership constraint. Certainly cvxpy already has a way to do this, but accessing gradients of composed affine operators provides nice way to users to debug their models. This could also help users who are trying to transition to a deployment that lacks a nice modeling interface (e.g. from cvxpy via python, to C or C++).

**Comment by SteveDiamond** [comment]
Ok I get it now. I 100% agree the ability to reduce/flatten affine expressions to coefficients should be more prominently exposed.

**Comment by rileyjmurray** [comment]
Wonderful! Yes a desire to access the coefficients of an affine expression is exactly how I noticed that grad behaved in its current way.

That said- there are many ways to expose the coefficients of affine expressions. Modifying grad for affine expressions is one way, but it might not be the best way. Modifying grad to return fully symbolic expressions would resolve this and provide some other interesting features (like more easily or transparently checking KKT conditions), but that would take more work. 

If the modifications are made *outside* of grad, then where would they go? I wouldn't want to have Expression objects be equipped with a "get_coefficients" function if that function only worked for affine expressions. 

Maybe this is something we should discuss in the phone call you were planning via whenisgood?

**Comment by SteveDiamond** [comment]
Yes let's discuss this in the dev meeting. The functionality you're asking for already exists actually and it's mostly a matter of documenting it.
